File: src/test/java/de/espend/idea/php/annotation/tests/doctrine/intention/DoctrineOrmRepositoryIntentionTest.java
package de.espend.idea.php.annotation.tests.doctrine.intention;

import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.psi.PsiElement;
import com.jetbrains.php.lang.PhpFileType;
import de.espend.idea.php.annotation.doctrine.intention.DoctrineOrmRepositoryIntention;
import de.espend.idea.php.annotation.tests.AnnotationLightCodeInsightFixtureTestCase;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see de.espend.idea.php.annotation.doctrine.intention.DoctrineOrmRepositoryIntention
 */
public class DoctrineOrmRepositoryIntentionTest extends AnnotationLightCodeInsightFixtureTestCase {

    public void setUp() throws Exception {
        super.setUp();
        myFixture.copyFileToProject("classes.php");
    }

    public String getTestDataPath() {
        return "src/test/java/de/espend/idea/php/annotation/tests/doctrine/intention/fixtures";
    }

    public void testThatRepositoryAnnotatorIsAvailableForAnnotation() {
        assertIntentionIsAvailable(PhpFileType.INSTANCE, "<?php\n" +
                "\n" +
                "use Doctrine\\ORM\\Mapping as ORM;\n" +
                "\n" +
                "/**\n" +
                " * @ORM\\Ent<caret>ity()\n" +
                " */\n" +
                "class Relation\n" +
                "{\n" +
                "}",
            "Add Doctrine repository"
        );
    }

    public void testThatRepositoryAnnotatorIsAvailableForAttribute() {
        assertIntentionIsAvailable(PhpFileType.INSTANCE, "<?php\n" +
                "\n" +
                "use Doctrine\\ORM\\Mapping as ORM;\n" +
                "\n" +
                "#[ORM\\Ent<caret>ity()]" +
                "class Relation\n" +
                "{\n" +
                "}",
            "Add Doctrine repository"
        );

        assertIntentionIsAvailable(PhpFileType.INSTANCE, "<?php\n" +
                "\n" +
                "use Doctrine\\ORM\\Mapping as ORM;\n" +
                "\n" +
                "#[ORM\\Entity(\"fo<caret>o\")]" +
                "class Relation\n" +
                "{\n" +
                "}",
            "Add Doctrine repository"
        );
    }

    public void testThatRepositoryAnnotatorForAttributeValueIsAvailable() {
        assertIntentionIsAvailable(PhpFileType.INSTANCE,"<?php\n" +
                "\n" +
                "use Doctrine\\ORM\\Mapping as ORM;\n" +
                "\n" +
                "/**\n" +
                " * @ORM\\Entity(<caret>)\n" +
                " */\n" +
                "class Relation\n" +
                "{\n" +
                "}",
            "Add Doctrine repository"
        );
    }

    public void testAttribute() {
        myFixture.configureByText("test.php", "<?php\n" +
            "\n" +
            "namespace App\\Entity;\n" +
            "\n" +
            "use App\\Entity\\Repository\\UserRepository;\n" +
            "use Doctrine\\ORM\\Mapping as ORM;\n" +
            "\n" +
            "#[ORM\\Ent<caret>ity()]\n" +
            "class User {}\n"
        );

        String s = invokeAndGetText();

        assertTrue(s.contains("use App\\Entity\\Repository\\UserRepository;"));
        assertTrue(s.contains("#[ORM\\Entity(repositoryClass: UserRepository::class)]"));
    }

    public void testAnnotation() {
        myFixture.configureByText("test.php", "<?php\n" +
            "\n" +
            "namespace App\\Entity;\n" +
            "\n" +
            "use Doctrine\\ORM\\Mapping as ORM;\n" +
            "\n" +
            "/**\n" +
            " * @ORM\\En<caret>tity()\n" +
            " */\n" +
            "class User {}\n"
        );

        String s = invokeAndGetText();

        assertTrue(s.contains("use App\\Entity\\Repository\\UserRepository;"));
        assertTrue(s.contains("@ORM\\Entity(repositoryClass=UserRepository::class)"));
    }

    private String invokeAndGetText() {
        PsiElement psiElement = myFixture.getFile().findElementAt(myFixture.getCaretOffset());

        WriteCommandAction.runWriteCommandAction(
            getProject(),
            () -> new DoctrineOrmRepositoryIntention().invoke(getProject(), getEditor(), psiElement)
        );

        return getEditor().getDocument().getText();
    }
}

File: src/test/kotlin/de/espend/idea/php/annotation/tests/doctrine/intention/DoctrineOrmRepositoryIntentionTest.kt
package de.espend.idea.php.annotation.tests.doctrine.intention

import com.intellij.openapi.command.WriteCommandAction
import com.jetbrains.php.lang.PhpFileType
import de.espend.idea.php.annotation.doctrine.intention.DoctrineOrmRepositoryIntention
import de.espend.idea.php.annotation.tests.AnnotationLightCodeInsightFixtureTestCase

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 * @see de.espend.idea.php.annotation.doctrine.intention.DoctrineOrmRepositoryIntention
 */
class DoctrineOrmRepositoryIntentionTest : AnnotationLightCodeInsightFixtureTestCase() {
    public override fun setUp() {
        super.setUp()
        myFixture.copyFileToProject("classes.php")
    }

    override fun getTestDataPath(): String {
        return "src/test/kotlin/de/espend/idea/php/annotation/tests/doctrine/intention/fixtures"
    }

    fun testThatRepositoryAnnotatorIsAvailableForAnnotation() {
        assertIntentionIsAvailable(
            PhpFileType.INSTANCE,
            """
                <?php

                use Doctrine\ORM\Mapping as ORM;

                /**
                 * @ORM\Ent<caret>ity()
                 */
                class Relation
                {
                }
            """.trimIndent(),
            "Add Doctrine repository",
        )
    }

    fun testThatRepositoryAnnotatorIsAvailableForAttribute() {
        assertIntentionIsAvailable(
            PhpFileType.INSTANCE,
            """
                <?php

                use Doctrine\ORM\Mapping as ORM;

                #[ORM\Ent<caret>ity()]class Relation
                {
                }
            """.trimIndent(),
            "Add Doctrine repository",
        )

        assertIntentionIsAvailable(
            PhpFileType.INSTANCE,
            """
                <?php

                use Doctrine\ORM\Mapping as ORM;

                #[ORM\Entity("fo<caret>o")]class Relation
                {
                }
            """.trimIndent(),
            "Add Doctrine repository",
        )
    }

    fun testThatRepositoryAnnotatorForAttributeValueIsAvailable() {
        assertIntentionIsAvailable(
            PhpFileType.INSTANCE,
            """
                <?php

                use Doctrine\ORM\Mapping as ORM;

                /**
                 * @ORM\Entity(<caret>)
                 */
                class Relation
                {
                }
            """.trimIndent(),
            "Add Doctrine repository",
        )
    }

    fun testAttribute() {
        myFixture.configureByText(
            "test.php",
            """
                <?php

                namespace App\Entity;

                use App\Entity\Repository\UserRepository;
                use Doctrine\ORM\Mapping as ORM;

                #[ORM\Ent<caret>ity()]
                class User {}
            """.trimIndent(),
        )

        val text = invokeAndGetText()
        assertTrue(text.contains("use App\\Entity\\Repository\\UserRepository;"))
        assertTrue(text.contains("#[ORM\\Entity(repositoryClass: UserRepository::class)]"))
    }

    fun testAnnotation() {
        myFixture.configureByText(
            "test.php",
            """
                <?php

                namespace App\Entity;

                use Doctrine\ORM\Mapping as ORM;

                /**
                 * @ORM\En<caret>tity()
                 */
                class User {}
            """.trimIndent(),
        )

        val text = invokeAndGetText()
        assertTrue(text.contains("use App\\Entity\\Repository\\UserRepository;"))
        assertTrue(text.contains("@ORM\\Entity(repositoryClass=UserRepository::class)"))
    }

    private fun invokeAndGetText(): String {
        val psiElement = myFixture.file.findElementAt(myFixture.caretOffset)

        WriteCommandAction.runWriteCommandAction(project) {
            DoctrineOrmRepositoryIntention().invoke(project, editor, psiElement!!)
        }

        return editor.document.text
    }
}
